tests/parser/syntax/test_msg_data.py

```diff
@@ -0,0 +1,118 @@
+import pytest
+from eth_tester.exceptions import TransactionFailed
+
+from vyper import compiler
+from vyper.exceptions import StateAccessViolation, SyntaxException
+
+
+def test_variable_assignment(get_contract, keccak):
+    code = """
+@external
+def foo() -> Bytes[4]:
+    bar: Bytes[4] = slice(msg.data, 0, 4)
+    return bar
+"""
+
+    contract = get_contract(code)
+
+    assert contract.foo() == bytes(keccak(text="foo()")[:4])
+
+
+def test_slicing_start_index_other_than_zero(get_contract):
+    code = """
+@external
+def foo(_value: uint256) -> uint256:
+    bar: Bytes[32] = slice(msg.data, 4, 32)
+    return convert(bar, uint256)
+"""
+
+    contract = get_contract(code)
+
+    assert contract.foo(42) == 42
+
+
+def test_get_full_calldata(get_contract, keccak, w3):
+    code = """
+@external
+def foo(bar: uint256) -> Bytes[36]:
+    data: Bytes[36] = slice(msg.data, 0, 36)
+    return data
+"""
+    contract = get_contract(code)
+
+    # 2fbebd38000000000000000000000000000000000000000000000000000000000000002a
+    method_id = keccak(text="foo(uint256)").hex()[2:10]  # 2fbebd38
+    encoded_42 = w3.toBytes(42).hex()  # 2a
+    expected_result = method_id + "00" * 31 + encoded_42
+
+    assert contract.foo(42).hex() == expected_result
+
+
+def test_get_len(get_contract):
+    code = """
+@external
+def foo(bar: uint256) -> uint256:
+    return len(msg.data)
+"""
+    contract = get_contract(code)
+
+    assert contract.foo(42) == 36
+
+
+fail_list = [
+    (
+        """
+@external
+def foo() -> Bytes[4]:
+    bar: Bytes[4] = msg.data
+    return bar
+    """,
+        SyntaxException,
+    ),
+    (
+        """
+@external
+def foo() -> Bytes[7]:
+    bar: Bytes[7] = concat(msg.data, 0xc0ffee)
+    return bar
+    """,
+        SyntaxException,
+    ),
+    (
+        """
+@external
```

Comment:
Also check that assignment to `bytes32` fails, as this type is the closest to `Bytes[32]` and there might be some unexpected weirdness in the type checking due to the overlap.